src/models.rs

```diff
@@ -217,6 +219,146 @@ impl<'w> Iterator for Jobs<'w> {
     }
 }
 
+/// Represents an execution Matrix within a Job.
+///
+/// This type implements [`Deref`] for [job::NormalJob::Strategy`], providing
+/// access to the underlying data model.
+#[derive(Clone)]
+pub(crate) struct Matrix<'w> {
+    inner: &'w LoE<job::Matrix>,
+}
+
+impl<'w> Deref for Matrix<'w> {
+    type Target = &'w LoE<job::Matrix>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<'w> TryFrom<&'w Job<'w>> for Matrix<'w> {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &'w Job<'w>) -> std::result::Result<Self, Self::Error> {
+        let workflow::Job::NormalJob(job) = value.deref() else {
+            bail!("job is not a normal job")
+        };
+
+        let Some(strategy) = &job.strategy else {
+            bail!("job does not define a strategy")
+        };
+
+        let Some(inner) = &strategy.matrix else {
+            bail!("job does not define a matrix")
+        };
```

Comment:
I gave a few tries on this one, tbh I could not figure out how to collapse it to a single `let` 🙈

Would `and_then` chaining be your take here?

Reply:
Ah yeah, I forgot this wouldn't work: `NormalJob` has a `Box<...>` in it, so we can't completely flatten these destructures (since `box_patterns` was never stabilized). Sorry for the wild goose chase here!

We can still reduce it slightly however; I'll push up what I'm thinking.

Reply:
7756e9e